### DISCUSSION TITLE
I can't generate an apk with Flash Builder 4.7 and AIRSDK 50.2.2.5

### DISCUSSION THREAD

**Discussion by MAC-Joyco** [airsdk/Adobe-Runtime-Support#2640] [General]
Hi!,

The last time I generated an apk without problems was with AIRSDK 33.1.1.575 but now I need the latest AIRSDK since I also need the latest ANE versions from Distriqt.

The error that Flash Builder shows me (which stops the process without generating anything) is: "Multiple Android build-tools folders exist, choosing the latest version (33.0.2) for packaging." 

but adt.log shows me the following:

`Target requested was apk-captive-runtime, but not using legacy APK so building via Android Studio
Creating device SDK handler - location passed in = C:\Users\Alejandro\AppData\Local\Android\Sdk
Version in APK = 50.2.2.5, version of ADT = 50.2.2.5
Version in APK = 50.2.2.5, version of ADT = 50.2.2.5
Version in APK = 50.2.2.5, version of ADT = 50.2.2.5
Version in APK = 50.2.2.5, version of ADT = 50.2.2.5
Gradle build where ADT is using JRE version 1.8.0_371
but where Gradle will be using the JDK from C:/Program Files/Java/jdk-11\bin\java.exe
Gradle build command:  C:/Program Files/Java/jdk-11\bin\java.exe -Xms1024m -Xmx1536m -Dorg.gradle.appname=gradlew -Dorg.gradle.jvmargs=-Xmx1536m -Duser.dir="C:\Users\ALEJAN~1\AppData\Local\Temp\1e9424b3-c55e-43e5-bc2e-1a2263fa003f" -cp "C:\Users\ALEJAN~1\AppData\Local\Temp\1e9424b3-c55e-43e5-bc2e-1a2263fa003f\gradle\wrapper\gradle-wrapper.jar";"C:\Program Files\Adobe\Adobe Flash Builder 4.7 (64 Bit)\eclipse\plugins\com.adobe.flash.compiler_4.7.0.349722\AIRSDK\lib\android\bin\gradle-cli.jar" org.gradle.wrapper.GradleWrapperMain :app:assembleRelease
..java.lang.Thread.getStackTrace(Unknown Source)
..com.adobe.air.DebugFile.printStack(DebugFile.java)
..com.adobe.air.ADTException.<init>(ADTException.java)
..com.adobe.air.apk.AABOutputStream.buildGradle(AABOutputStream.java)
..com.adobe.air.apk.AABPackager.renameOutput(AABPackager.java)
..com.adobe.air.ApplicationPackager.createPackage(ApplicationPackager.java)
..com.adobe.air.apk.AABPackager.createPackage(AABPackager.java)
..com.adobe.air.ADT.parseArgsAndGo(ADT.java)
..com.adobe.air.ADT.run(ADT.java)
..com.adobe.air.ADT.main(ADT.java)
Stack trace:
Unexpected failure: Unable to run java: com.adobe.air.ADTException: gradle tool failed: Warning: This version only understands SDK XML versions up to 2 but an SDK XML file of version 3 was encountered. This can happen if you use versions of Android Studio and the command-line tools that were released at different times.
`

It seems that the problem FB shows me had little to do with the real problem, could the problem be that adt is using JRE 1.8.0_371 while Gradle uses jdk-11?

I think Gradle uses JDK 11 because I included JAVA_HOME=C:\\Program Files\\Java\\jdk-11 in adt.cfg, while adt uses the internal java version of Flash Builder, which is 1.8.0_371

if my assumption is correct, I would need to know if there is any way to update the java version that Flash Builder uses to version 11. I have updated java in Flash Builder other times by simply replacing the jre folder that comes inside the Flash Builder application folder, but java jdk 11 no longer comes with jre. I tried creating my own jre with jlink from jdk 11, but it didn't work and Flash Builder won't initialize. I also tried to put the jdk 11 as if it was the jre and it doesn't work either.

I'm afraid I will have to migrate the project to another IDE, I would like to avoid that step for the moment as I don't have too much time, but if I had to migrate to another IDE, which one do you recommend?

Thanks in advance!

**Comment by ajwfrost** [comment]
Hi
Is there anything more shown in the log? There is a line "gradle tool failed" but then no actual error listed, I would expect there to be quite a bit of text after that, and somewhere within this is the actual error...

ADT is fine to run with Java 7 or 8, but then when we trigger Gradle, we check and yes would use JDK 11 based on that value. And this is what's actually running:
```
C:/Program Files/Java/jdk-11\bin\java.exe -Xms1024m -Xmx1536m -Dorg.gradle.appname=gradlew -Dorg.gradle.jvmargs=-Xmx1536m -Duser.dir="C:\Users\ALEJAN~1\AppData\Local\Temp\1e9424b3-c55e-43e5-bc2e-1a2263fa003f" -cp "C:\Users\ALEJAN~1\AppData\Local\Temp\1e9424b3-c55e-43e5-bc2e-1a2263fa003f\gradle\wrapper\gradle-wrapper.jar";"C:\Program Files\Adobe\Adobe Flash Builder 4.7 (64 Bit)\eclipse\plugins\com.adobe.flash.compiler_4.7.0.349722\AIRSDK\lib\android\bin\gradle-cli.jar" 
```

So I don't think you would need to update the JDK/JRE within Flash Builder, but it would be good to find out why the Gradle project is failing to build..

**Reply by MAC-Joyco** [reply]
Thanks!, you were right, fortunately that wasn't the problem. There was a duplicate activity in the manifest and I also updated the Android SDK and that fixed the problem.

**Comment by jason68000** [comment]
Hello! How do you generate the adt.log? I tried the adt.cfg file but no luck! Thanks, Jason

**Reply by ajwfrost** [reply]
Simplest is to use the AIR SDK Manager .. there's a "troubleshooting" page which will automatically display the log if you run ADT whilst having that view open.

Or you can enable a log file generation using the "adt.cfg" file which is now stored in a folder called ".airsdk", under your home folder (so e.g. c:\users\username\.airsdk\adt.cfg on Windows, or /Users/username/.airsdk/adt.cfg on macOS). If you add a line `DebugOut=true` this should then result in generation of a file "adt.log" in the home folder when you next run ADT.

**Reply by jason68000** [reply]
Go it, thanks! 

Another quick question: Do you know how to configure adt.cfg to use the android SDK?

I've added:

AndroidPlatformSDK=C:/Users/Jason/AppData/Local/Android/Sdk/build-tools/36.0.0      # Incorrect?
AndroidPlatformSDK=C:/Users/Jason/AppData/Local/Android\Sdk                                    # Correct?

If the second AndroidPlatformSDK is correct do I also need to set the SDK version number in the adt.cfg file?
And do I use forward slashes?

I can't seem to get of the legacy build - ie, I don't think anything I'm doing here is changing my build when I compare adt.log it's always the same, regardless of what/if I assign to AndroidPlatformSDK  in the adt.cfg.

**Reply by marchbold** [reply]
Try the AIR SDK manager and use the path to the android SDK that's listed in android studio .You have the path to the build tools there which is different

**Reply by marchbold** [reply]
Oh sorry your second entry looks correct just needs formatting changes

**Comment by jason68000** [comment]
Hello everybody on this group!

I'm also trying to generate an APK through Flash Builder 4.7.. I just can't see to get off the legacy build system. I thought FB 4.7 might be doing it's own thing and cannot be changed but from the adt.log above it looks like you're able to use ANE versions from Distriqt and build correctly.

You can read my discussion here from another post:
https://github.com/airsdk/Adobe-Runtime-Support/discussions/3477

Essentially I've tried everything on the list but I just can't shake of the legacy build. Here's a snipped of my adt.log:

Called by: C:\Program Files\Adobe\Adobe Flash Builder 4.7 (64 Bit)\sdks\4.6.0\lib\adt.jar
APK Packager entry point
AAPT command line: C:\Program Files\Adobe\Adobe Flash Builder 4.7 (64 Bit)\sdks\4.6.0\lib\android\bin\aapt.exe package -z -u -J XXXX...

When I compare this with the new build (just posted from FliplineStudios) you can see it looks much more modern:

Called by: C:\dev\AIRSDK_Windows51_1_3_10\lib\adt.jar
ADT 51.1.3.10 called with: -package -target apk-captive-runtime -arch armv7 -storetype pkcs12 [Delete]
Target requested was apk-captive-runtime, but not using legacy APK so building via Android Studio
Creating device SDK handler - location passed in = C:\Users\Matt\AppData\Local\Android\Sdk
Adding gradle dependency: implementation "com.android.billingclient:billing:6.2.1"
Adding gradle dependency: implementation 'org.jetbrains:annotations:13.0'

MAC-Joyco did you go through a similar set of problems when using FB 4,.7 to export your APK?

THanks!
Jason

**Reply by ajwfrost** [reply]
Your adt.log file should give us the details of why it's choosing the legacy build .. it might be that it can't find an appropriate JDK installation, or the platformsdk path, or your app descriptor namespace version may be too old (33.1 onwards supports the new gradle/android studio mechanism).

But if all you are seeing in adt.log is the "Called by" line followed directly by "APK Packager entry point" then it looks like Flash Builder isn't actually calling the main ADT class entry point which is where all this logic goes on... Actually yes, I just checked the code, this is calling the legacy class directly (APK packager is the old way of doing things, "AAB Packager" is the class that uses Gradle to build things (originally just AAB then we switched the full Android build system over to this).

I think you can import Flash Builder projects into VSCode using the BowlerHat AS3/MXMXL extension, or into Moonshine IDE .. might be worth taking the plunge! The Flash Builder team had access into the internals of ADT so had all these shortcut mechanisms which will no longer work :-(

thanks
